### PR TITLE
WIP: Begin contributing document

### LINKED ISSUE #1376: Create documentation on how to get started as a contributor

### PR DESCRIPTION
**Pull request by odow**
Closes #1376

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+Before you do however, make sure you read the following before posting a new bug
+report:
+
+- If you have a question or are unsure if the behavior you're experiencing is a
+  bug, please search or post to our [Discourse site](https://discourse.julialang.org/c/domain/opt).
+  Questions posted to Discourse have broader visibility and are likely to be
+  answered more quickly than Github issues.
+
+- If you're experiencing a bug that is solver-specific (e.g., it only happens
```
Comment (blegat) [inline CONTRIBUTING.md]: Not a native but I have read that "you are" should be used rather than "you're" in written English

**Comment by mohamed82008** [comment]
Contributing examples and to the documentation?

**Comment by mohamed82008** [comment]
What happens after opening the issue? What if the potential contributor is willing to fix this issue by a PR. There needs to be some recommendations for making a PR. I can think of the following:

1. Follow the style guide,
2. Add tests and documentation for any changes or new features, 
3. Discuss any big changes in the issue before spending too much time on it to test the waters first and see if other contributors are fine with these changes or not,
4. Be patient and polite; you may get ALOT of comments on your PR!

Also, I see no mention of the Gitter channel, is that to cut down the noise there or a general disinterest in making Gitter the "official" go to channel?

**Comment by mlubin** [comment]
> Follow the style guide,
> Add tests and documentation for any changes or new features,
> Discuss any big changes in the issue before spending too much time on it to test the waters first and see if other contributors are fine with these changes or not,
> Be patient and polite; you may get ALOT of comments on your PR!

These are all good points that should be included.

> Also, I see no mention of the Gitter channel, is that to cut down the noise there or a general disinterest in making Gitter the "official" go to channel?

For me, the Gitter is the "official" channel for development questions (it's listed on the README). So yes, it should be mentioned.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+Github issue (more on this below).
+
+Aside from asking questions, the easiest way you can contribute to JuMP is to
+help answer questions on Discourse!
+
+## File a bug report
+
+Another way to contribute to JuMP is to file [bug reports](https://github.com/JuliaOpt/JuMP.jl/issues/new?template=bug_report.md).
+
+Before you do however, make sure you read the following before posting a new bug
```
Comment (mlubin) [inline CONTRIBUTING.md]: This repeats the same points from the bug report template. We can refer to that instead.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+If your change is larger, or touches multiple files, you will need to make the
+change locally and then use Git to submit a pull request. (See [Contribute code to JuMP](#Contribute code to JuMP)
+below for more on this.)
+
+If you need any help, come join our [Gitter](https://gitter.im/JuliaOpt/JuMP-dev)
+channel and we will walk you through the process.
+
+## Write examples
+
+Another way to contribute to JuMP is to add examples. The easiest way to do this
```
Comment (mlubin) [inline CONTRIBUTING.md]: The structure for examples is a bit up in the air now (https://github.com/JuliaOpt/JuMP.jl/issues/1175), so I'm not sure if we're ready to accept new ones at this point.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+channel and we will walk you through the process.
+
+Once you are familiar with Git, the workflow for contributing code to JuMP is
+along the lines of the following:
+1. Find an [open issue](https://github.com/JuliaOpt/JuMP.jl/issues) (or open a
+    new one) for the problem you want to solve
+2. Discuss (in the issue, or on [Gitter](https://gitter.im/JuliaOpt/JuMP-dev))
+_before_ spending too much time on it to test the waters first and see if other
+contributors are fine with these changes or not
+3. Make your changes locally. (The Julia manual has a [guide ](https://docs.julialang.org/en/stable/manual/packages/#Making-changes-to-an-existing-package-1)
```
Comment (mlubin) [inline CONTRIBUTING.md]: Nit: extra space in `[guide ]`

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+on how to do this.) Make sure you:
+    - Follow the [style guide](http://www.juliaopt.org/JuMP.jl/latest/style.html)
+    - Add tests and documentation for any changes or new features
+4. Submit a pull request to Github
+5. Update your pull request, responding to any comments
+
+Remember to be patient and polite; you may get _alot_ of comments on your PR!
```
Comment (mlubin) [inline CONTRIBUTING.md]: Nit: "alot" -> "a lot"

**Comment by mlubin** [comment]
GitHub [suggests](https://help.github.com/articles/setting-guidelines-for-repository-contributors/) linking to the code of conduct.

**Comment by mlubin** [comment]
Also cross link from https://github.com/JuliaOpt/JuMP.jl/blob/master/CODE_OF_CONDUCT.md#get-involved (which currently says "We haven't yet put together a guide for new contributors ...").

**Comment by odow** [comment]
Yeah I just copied [Atom](https://github.com/atom/atom/blob/master/CONTRIBUTING.md)

**Comment by mlubin** [comment]
LGTM now. Let's get a few more eyeballs.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+before posting. You can also find a copy of that info [here](https://github.com/JuliaOpt/JuMP.jl/blob/master/.github/ISSUE_TEMPLATE/bug_report.md).)
+
+## Contribute code to JuMP
+
+Finally, you can also contribute code to JuMP!
+
+If you don't have experience with Git, Github, and Julia development, the first
+steps can be a little daunting. However, there are lots of tutorials available
+online (such as [this](http://try.github.io/), [this](https://guides.github.com/activities/hello-world/),
+and [this](https://docs.julialang.org/en/stable/manual/packages/#Making-changes-to-an-existing-package-1)).
```
Comment (blegat) [inline CONTRIBUTING.md]: Maybe also Git Real of code school and the git-book

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+If you need any help, come join our [Gitter](https://gitter.im/JuliaOpt/JuMP-dev)
+channel and we will walk you through the process.
+
+Once you are familiar with Git, the workflow for contributing code to JuMP is
+along the lines of the following:
+1. Find an [open issue](https://github.com/JuliaOpt/JuMP.jl/issues) (or open a
+    new one) for the problem you want to solve
+2. Discuss (in the issue, or on [Gitter](https://gitter.im/JuliaOpt/JuMP-dev))
+_before_ spending too much time on it to test the waters first and see if other
```
Comment (blegat) [inline CONTRIBUTING.md]: Missing indent

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+contributors are fine with these changes or not
+3. Make your changes locally. (The Julia manual has a [guide](https://docs.julialang.org/en/stable/manual/packages/#Making-changes-to-an-existing-package-1)
+on how to do this.) Make sure you:
```
Comment (blegat) [inline CONTRIBUTING.md]: Missing indent

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+The docs are written in Markdown and are built using [Documenter.jl](https://github.com/JuliaDocs/Documenter.jl).
+You can find the source of all the docs [here](https://github.com/JuliaOpt/JuMP.jl/tree/master/docs).
+
+If your change is small (like fixing typos, or one or two sentence corrections),
+the easiest way to do this is via Github's online editor. (Github has
+[help](https://help.github.com/articles/editing-files-in-another-user-s-repository/)
+on how to do this.)
+
+If your change is larger, or touches multiple files, you will need to make the
+change locally and then use Git to submit a pull request. (See [Contribute code to JuMP](#Contribute code to JuMP)
```
Comment (mohamed82008) [inline CONTRIBUTING.md]: This shows up in a funny way in the file, not sure what the fix is if any is needed.

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+If you need any help, come join our [Gitter](https://gitter.im/JuliaOpt/JuMP-dev)
+channel and we will walk you through the process.
+
+Once you are familiar with Git, the workflow for contributing code to JuMP is
+along the lines of the following:
+1. Find an [open issue](https://github.com/JuliaOpt/JuMP.jl/issues) (or open a
+    new one) for the problem you want to solve
+2. Discuss (in the issue, or on [Gitter](https://gitter.im/JuliaOpt/JuMP-dev))
+    _before_ spending too much time on it to test the waters first and see if
+    other contributors are fine with these changes or not
```
Comment (mohamed82008) [inline CONTRIBUTING.md]: Discuss what and which changes?

**Comment by mlubin** [comment]
Seems good enough to merge. Can be tweaked further as needed.

**Comment by mlubin** [comment]
Thanks @odow !

**Pull request merged**